lib/auth/auth_with_roles.go

```diff
@@ -5946,44 +5930,100 @@ func (a *ServerWithRoles) ReplaceRemoteLocks(ctx context.Context, clusterName st
 // channel if one is encountered. Otherwise the event channel is closed when the stream ends.
 // The event channel is not closed on error to prevent race conditions in downstream select statements.
 func (a *ServerWithRoles) StreamSessionEvents(ctx context.Context, sessionID session.ID, startIndex int64) (chan apievents.AuditEvent, chan error) {
-	createErrorChannel := func(err error) (chan apievents.AuditEvent, chan error) {
+	// StreamSessionEvents can be called internally, and when that happens we
+	// don't want to emit an event or check permissions.
+	if err := a.localServerAction(); err == nil {
+		return a.alog.StreamSessionEvents(ctx, sessionID, startIndex)
+	}
+
+	if err := a.actionForKindSession(ctx, apidefaults.Namespace, sessionID); err != nil {
+		c, e := make(chan apievents.AuditEvent), make(chan error, 1)
+		e <- trace.Wrap(err)
+		return c, e
+	}
+
+	eventsCh, errCh := a.alog.StreamSessionEvents(ctx, sessionID, startIndex)
```

Comment:
we will miss the first event when startIndex is non-zero. How complex would it be if we had to change the StreamSessionEvents? This is on auth so no need to change client and gRPC

Reply:
The events won't be lost. The only difference is that we won't be able to set the session type (because the event type differs from the ones we're checking). In this case, the value would be unknown (empty string).